crates/cranelift/src/compiler.rs

```diff
@@ -166,19 +166,86 @@ impl wasmtime_environ::Compiler for Compiler {
         self.save_translator(func_translator);
 
         let mut code_buf: Vec<u8> = Vec::new();
-        let mut reloc_sink = RelocSink::new();
-        let mut trap_sink = TrapSink::new();
-        let mut stack_map_sink = StackMapSink::default();
         context
-            .compile_and_emit(
-                isa,
-                &mut code_buf,
-                &mut reloc_sink,
-                &mut trap_sink,
-                &mut stack_map_sink,
-            )
+            .compile_and_emit(isa, &mut code_buf)
             .map_err(|error| CompileError::Codegen(pretty_error(&context.func, error)))?;
 
+        let result = context.mach_compile_result.as_ref().unwrap();
+
+        let mut func_relocs = Vec::new();
+        for &MachReloc {
+            offset,
+            srcloc: _,
+            kind,
+            ref name,
+            addend,
+        } in result.buffer.relocs()
+        {
+            let reloc_target = if let ExternalName::User { namespace, index } = *name {
+                debug_assert_eq!(namespace, 0);
+                RelocationTarget::UserFunc(FuncIndex::from_u32(index))
+            } else if let ExternalName::LibCall(libcall) = *name {
+                RelocationTarget::LibCall(libcall)
+            } else {
+                panic!("unrecognized external name")
+            };
+            func_relocs.push(Relocation {
+                reloc: kind,
+                reloc_target,
+                offset,
+                addend,
+            });
+        }
+
+        let mut traps = Vec::new();
+        for &MachTrap {
+            offset,
+            srcloc: _,
+            code,
+        } in result.buffer.traps()
+        {
```

Comment:
Likewise here and below for stackmaps: factor out into separate conversion functions.